contrib/PyCuAmpcor/src/cuEstimateStats.cu

```diff
@@ -58,72 +90,73 @@ __global__ void cudaKernel_estimateVar(const float* corrBatchRaw, const int NX,
     // Preparation.
     int px = maxloc[idxImage].x;
     int py = maxloc[idxImage].y;
-    float peak = maxval[idxImage];
+    real_type peak = maxval[idxImage];
 
     // Check if maxval is on the margin.
-    if (px-1 < 0 || py-1 <0 || px + 1 >=NX || py+1 >=NY)  {
+    if (px-distance < 0 || py-distance <0 || px + distance >=NX || py+distance >=NY)  {
 
-        covValue[idxImage] = make_float3(99.0, 99.0, 0.0);
+        covValue[idxImage] = make_real3(99.0, 99.0, 0.0);
 
     }
     else {
         int offset = NX * NY * idxImage;
-        int idx00 = offset + (px - 1) * NY + py - 1;
-        int idx01 = offset + (px - 1) * NY + py    ;
-        int idx02 = offset + (px - 1) * NY + py + 1;
-        int idx10 = offset + (px    ) * NY + py - 1;
+        int idx00 = offset + (px - distance) * NY + py - distance;
+        int idx01 = offset + (px - distance) * NY + py    ;
+        int idx02 = offset + (px - distance) * NY + py + distance;
+        int idx10 = offset + (px    ) * NY + py - distance;
         int idx11 = offset + (px    ) * NY + py    ;
-        int idx12 = offset + (px    ) * NY + py + 1;
-        int idx20 = offset + (px + 1) * NY + py - 1;
-        int idx21 = offset + (px + 1) * NY + py    ;
-        int idx22 = offset + (px + 1) * NY + py + 1;
+        int idx12 = offset + (px    ) * NY + py + distance;
+        int idx20 = offset + (px + distance) * NY + py - distance;
+        int idx21 = offset + (px + distance) * NY + py    ;
+        int idx22 = offset + (px + distance) * NY + py + distance;
 
         // second-order derivatives
-        float dxx = - ( corrBatchRaw[idx21] + corrBatchRaw[idx01] - 2.0*corrBatchRaw[idx11] );
-        float dyy = - ( corrBatchRaw[idx12] + corrBatchRaw[idx10] - 2.0*corrBatchRaw[idx11] ) ;
-        float dxy = ( corrBatchRaw[idx22] + corrBatchRaw[idx00] - corrBatchRaw[idx20] - corrBatchRaw[idx02] ) *0.25;
+        real_type dxx = - ( corrBatchRaw[idx21] + corrBatchRaw[idx01] - 2.0*corrBatchRaw[idx11] );
+        real_type dyy = - ( corrBatchRaw[idx12] + corrBatchRaw[idx10] - 2.0*corrBatchRaw[idx11] ) ;
+        real_type dxy = ( corrBatchRaw[idx22] + corrBatchRaw[idx00] - corrBatchRaw[idx20] - corrBatchRaw[idx02] ) *0.25;
 
-        float n2 = fmaxf(1.0 - peak, 0.0);
+        real_type n2 = max(1.0 - peak, 0.0);
 
         dxx = dxx * templateSize;
         dyy = dyy * templateSize;
         dxy = dxy * templateSize;
 
-        float n4 = n2*n2;
+        real_type n4 = n2*n2;
         n2 = n2 * 2;
         n4 = n4 * 0.5 * templateSize;
 
-        float u = dxy * dxy - dxx * dyy;
-        float u2 = u*u;
+        real_type u = dxy * dxy - dxx * dyy;
+        real_type u2 = u*u;
 
         // if the Gaussian curvature is too small
         if (fabsf(u) < 1e-2) {
-            covValue[idxImage] = make_float3(99.0, 99.0, 0.0);
+            covValue[idxImage] = make_real3(99.0, 99.0, 0.0);
```

Comment:
To my understanding, if the Gaussian curvature is too small (less than 0.01) then we assign a `noData` value of 99. Why we do this and where is the threshold of 0.01 come from?

Reply:
This follows the ROIPAC. We could make it a user configurable parameter.